### ISSUE TITLE
Unable to batch push data on the index

### ISSUE THREAD

**Issue by nfrankel** [algolia/jekyll-algolia#38]
Hello,

Just found out about Algolia and its Jekyll plugin.

However, when I run the command:

```
ALGOLIA_API_KEY='secret' bundle exec jekyll algolia --trace
```

I stumble upon the following issue:

```
Extracting records...                                                            
Updating records in index blog.frankel.ch...                                     
Records to delete: 33                                                            
Records to add:    57                                                            
bundler: failed to load command: jekyll (/Users/i303869/.rbenv/versions/2.4.2/bin/jekyll)
NoMethodError: undefined method `[]' for nil:NilClass
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/jekyll-algolia-1.1.0/lib/jekyll/algolia/error_handler.rb:198:in `record_too_big?'
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/jekyll-algolia-1.1.0/lib/jekyll/algolia/error_handler.rb:58:in `block in identify'
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/jekyll-algolia-1.1.0/lib/jekyll/algolia/error_handler.rb:57:in `each'
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/jekyll-algolia-1.1.0/lib/jekyll/algolia/error_handler.rb:57:in `identify'
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/jekyll-algolia-1.1.0/lib/jekyll/algolia/error_handler.rb:23:in `stop'
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/jekyll-algolia-1.1.0/lib/jekyll/algolia/indexer.rb:154:in `rescue in block in update_records'
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/jekyll-algolia-1.1.0/lib/jekyll/algolia/indexer.rb:151:in `block in update_records'
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/jekyll-algolia-1.1.0/lib/jekyll/algolia/indexer.rb:150:in `each'
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/jekyll-algolia-1.1.0/lib/jekyll/algolia/indexer.rb:150:in `each_slice'
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/jekyll-algolia-1.1.0/lib/jekyll/algolia/indexer.rb:150:in `update_records'
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/jekyll-algolia-1.1.0/lib/jekyll/algolia/indexer.rb:192:in `run'
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/jekyll-algolia-1.1.0/lib/jekyll-algolia.rb:119:in `write'
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/jekyll-3.7.2/lib/jekyll/site.rb:75:in `process'
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/jekyll-algolia-1.1.0/lib/jekyll-algolia.rb:57:in `run'
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/jekyll-algolia-1.1.0/lib/jekyll/commands/algolia.rb:42:in `block (2 levels) in init_with_program'
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `block in execute'
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `each'
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `execute'
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/mercenary-0.3.6/lib/mercenary/program.rb:42:in `go'
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/mercenary-0.3.6/lib/mercenary.rb:19:in `program'
  /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/jekyll-3.7.2/exe/jekyll:15:in `<top (required)>'
  /Users/i303869/.rbenv/versions/2.4.2/bin/jekyll:23:in `load'
  /Users/i303869/.rbenv/versions/2.4.2/bin/jekyll:23:in `<top (required)>'
```

Interestingly enough, if I remove all the posts (but not the pages), or use the example repo, it works.

Do you have any way I could debug further? Any hint/help appreciated.

My repo is hosted on a private Gitlab repo, but should the need be, I could give you access rights.

**Comment by DirtyF** [comment]
>   /Users/i303869/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/jekyll-algolia-1.1.0/lib/jekyll/algolia/error_handler.rb:198:in `record_too_big?'

Records are not expected to go beyond 10KB of minified JSON. How did you configure the plugin?

**Comment by nfrankel** [comment]
Thanks for your very fast feedback!

I left pretty much the default configuration:

```
algolia:
  application_id: xxx
  index_name: yyy
  indexing_batch_size: 500
  extensions_to_index:
    - html
    - md
    - adoc
```

In the end, I have ~6000 records to add.

**Comment by pixelastic** [comment]
Thanks for the report.

I stumbled upon a similar issue yesterday. What I think is going on is that the plugin is trying to push a record that is too big for Algolia (more than 10Kb). The plugin **should** handle this gracefully and let you know about the record that is indeed causing issues, but this part of the code is failing.

Now, why is your record that big? This is often caused either by a malformed HTML (unclosed tags, causing the HTML parser to grab too much content), or by recursive rendering in your pages (calling `{{ content }}` from a page, forcing it to re-render itself inside itself, inception-style).

I'll release a new version of the plugin soon, with better error handling that might help you pinpoint the issue, but if you can grant me access to your GitLab, or give me a  smaller reproducible test case I'd be happy to have a look.

**Comment by nfrankel** [comment]
> I'll release a new version of the plugin soon, with better error handling that might help you pinpoint the issue

That's exactly what I'm requesting. Many thanks.

Now, if you're kind enough to have a look at my own repo, that would be much appreciated: I give you access right now.

**Comment by pixelastic** [comment]
I just pulled your branch and tried a `jekyll algolia`. I managed to push ~600 records to my index with no error thrown.

But I realized that re-running the command was actually deleting all records and re-adding them. I'm going to get that fixed (I think it has something to do with the `adoc` converted). Maybe both bugs are related.

In the meantime, could you try to update to the latest v1.1.1. It should have better error handling of the error you were having. And to answer your question, records are pushed from memory and not written on disk. The new feature in `v1.1.1` will write a log file in your source directory containing the content of the record that was refused by Algolia. This should help you troubleshoot the origin of the issue.

**Comment by nfrankel** [comment]
Now, the console tells me what post is guilty and the log pinpoints it exactly. That's a huge step forward, many thanks for that!

The next issue is that the plugin is making some assumptions regarding excerpts.
By default, if no excerpt has been manually defined, it takes the whole content.
I know Jenkins makes the same assumption, but it's the reason why it fails.

(My blog has currently +400 posts so setting an excerpt on each post is not  an option, at least in the short term).

**Comment by pixelastic** [comment]
Jekyll actually has an internal way of extracting excerpts from pages/posts. The plugin is just calling that method and saving it in the record. I think that it got confused on how to extract it from asciidoc files (worked find on my tests on `html` and `md` files).

The other issue I mentioned above about pushing the same records over and over again was also due to an asciidoc issue (`jekyll-asciidoc` is adding a `document` key to each record that is causing the diff algorithm to fail). I have a fix for that in my local branch, I'm currently writing tests for it.

Would you mind adding a post to the `feature/search` branch where the `excerpt` is wrongly taking the whole content? I'd like to see if I can work around that directly inside the plugin, or if maybe that's an issue that should be fixed in `jekyll-asciidoc` directly. If I can't fix it from `jekyll-algolia`, I think you can fix it  with a [custom hook](https://community.algolia.com/jekyll-algolia/hooks.html#before-indexing-each) that will just delete the `excerpt` keys before pushing them.

**Comment by nfrankel** [comment]
A hook is definitely a terrific idea! This way, I could just shorten the field if the need be.

The post which creates trouble is `2017-11-12-oop-compatible-enterprise-context.adoc` and is already part of the branch.

**Comment by jdwallace12** [comment]
This is a timely discussion, I am also experiencing similar issue.  I updated the the jekyll algolia gem to 1.1.1  which gives the name of the file that errors on (super nice!).  It is a collection page that uses {{ content }} that renders the content into a layout.  I am trying to just exclude it but having trouble doing so.
  
<img width="957" alt="screen shot 2018-02-14 at 10 36 28 am" src="https://user-images.githubusercontent.com/3752170/36212806-f4204db2-1172-11e8-85e9-6a6d0133cbe5.png">

would this be the proper way to exclude a file that is within a collection or is it possible to do it in the yaml front matter of the page in question?

<img width="795" alt="screen shot 2018-02-14 at 10 37 28 am" src="https://user-images.githubusercontent.com/3752170/36212859-1488557c-1173-11e8-98bf-de850c1e1274.png">

**Comment by pixelastic** [comment]
@jdwallace12: The `files_to_exclude` should contain path to the original files (before transformation by jekyll), relative to your Jekyll soure. In your case that might be something like `./_resources/whitepapers/six-steps-for-it-troubleshooting-at-remote-locations.yml` or something similar.

If this does not work, would you mind opening a new issue and if possible adding a link to a git repo so I can reproduce the issue locally? 6Kb for an excerpt seems too big. I'm starting to wonder if adding excerpts to records is actually worth it :)

@nfrankel I fixed the "double pushing of records" issue (available on `develop` only for now). I'm going to have a look at your excerpt now :)

**Comment by nfrankel** [comment]
As far as I understand, a post will get spit into different records, but each of those records will get the same excerpt data. Am I right?

**Comment by pixelastic** [comment]
Yep, that's the way it works. 

Each page/post will be split into several records (one per paragraph), but each will share the page metadata.

I just re-tested on my side, and I can correctly index your post, with no limit reached. Could you let me know which version of Ruby and Jekyll you're using?

**Comment by nfrankel** [comment]
* Ruby: ruby 2.4.2p198
* jekyll: 3.7.2

**Comment by pixelastic** [comment]
We have the exact same versions. Could you share the actual JSON of the record that fails?

**Comment by nfrankel** [comment]
```
{
  "html": "<dd>\n<p>There are two ways to weave the aspect into the , either compile-time weaving, or load-time weaving.\nI choose the the later is much easier to configure.\nIt’s achieved through a standard Java agent.</p>\n<ol class=\"arabic\">\n<li>\n<span class=\"primary\">First, it needs to be added as a runtime dependency in the POM:</span><div class=\"listingblock\">\n<div class=\"content\">\n<pre class=\"highlight\"><code class=\"language-xml\" data-lang=\"xml\"><span id=\"L1\" class=\"line\"><span class=\"nt\">&lt;dependency&gt;</span></span>\n<span id=\"L2\" class=\"line\">    <span class=\"nt\">&lt;groupId&gt;</span>org.springframework<span class=\"nt\">&lt;/groupId&gt;</span></span>\n<span id=\"L3\" class=\"line\">    <span class=\"nt\">&lt;artifactId&gt;</span>spring-agent<span class=\"nt\">&lt;/artifactId&gt;</span></span>\n<span id=\"L4\" class=\"line\">    <span class=\"nt\">&lt;version&gt;</span>2.5.6<span class=\"nt\">&lt;/version&gt;</span></span>\n<span id=\"L5\" class=\"line\">    <span class=\"nt\">&lt;scope&gt;</span>runtime<span class=\"nt\">&lt;/scope&gt;</span></span>\n<span id=\"L6\" class=\"line\"><span class=\"nt\">&lt;/dependency&gt;</span></span></code></pre>\n</div>\n</div>\n</li>\n<li>\n<span class=\"primary\">Then, the Spring Boot plugin must be configured with the agent:</span><div class=\"listingblock\">\n<div class=\"content\">\n<pre class=\"highlight\"><code class=\"language-xml\" data-lang=\"xml\"><span id=\"L1\" class=\"line\"><span class=\"nt\">&lt;plugin&gt;</span></span>\n<span id=\"L2\" class=\"line\">    <span class=\"nt\">&lt;groupId&gt;</span>org.springframework.boot<span class=\"nt\">&lt;/groupId&gt;</span></span>\n<span id=\"L3\" class=\"line\">    <span class=\"nt\">&lt;artifactId&gt;</span>spring-boot-maven-plugin<span class=\"nt\">&lt;/artifactId&gt;</span></span>\n<span id=\"L4\" class=\"line\">    <span class=\"nt\">&lt;configuration&gt;</span></span>\n<span id=\"L5\" class=\"line\">        <span class=\"nt\">&lt;agent&gt;</span>${settings.localRepository}/org/springframework/spring-agent/2.5.6/spring-agent-2.5.6.jar<span class=\"nt\">&lt;/agent&gt;</span></span>\n<span id=\"L6\" class=\"line\">    <span class=\"nt\">&lt;/configuration&gt;</span></span>\n<span id=\"L7\" class=\"line\"><span class=\"nt\">&lt;/plugin&gt;</span></span></code></pre>\n</div>\n</div>\n</li>\n<li>\n<span class=\"primary\">Finally, the application must be configured accordingly:</span><div class=\"listingblock\">\n<div class=\"content\">\n<pre class=\"highlight\"><code class=\"language-kotlin\" data-lang=\"kotlin\"><span id=\"L1\" class=\"line\"><span class=\"n\">@EnableLoadTimeWeaving</span></span>\n<span id=\"L2\" class=\"line\"><span class=\"kd\">class</span> <span class=\"nc\">OopspringApplication</span></span></code></pre>\n</div>\n</div>\n</li>\n</ol>\n</dd>",
  "content": "\nThere are two ways to weave the aspect into the , either compile-time weaving, or load-time weaving.\nI choose the the later is much easier to configure.\nIt’s achieved through a standard Java agent.\nFirst, it needs to be added as a runtime dependency in the POM:\n\n<dependency>\n    <groupId>org.springframework</groupId>\n    <artifactId>spring-agent</artifactId>\n    <version>2.5.6</version>\n    <scope>runtime</scope>\n</dependency>\n\nThen, the Spring Boot plugin must be configured with the agent:\n\n<plugin>\n    <groupId>org.springframework.boot</groupId>\n    <artifactId>spring-boot-maven-plugin</artifactId>\n    <configuration>\n        <agent>${settings.localRepository}/org/springframework/spring-agent/2.5.6/spring-agent-2.5.6.jar</agent>\n    </configuration>\n</plugin>\n\nFinally, the application must be configured accordingly:\n\n@EnableLoadTimeWeaving\nclass OopspringApplication\n\n\n",
  "tag_name": "dd",
  "hierarchy": {
    "lvl0": null,
    "lvl1": "The final touch",
    "lvl2": null,
    "lvl3": null,
    "lvl4": null,
    "lvl5": null,
    "lvl6": null
  },
  "anchor": "the-final-touch",
  "weight": {
    "position": 23,
    "heading": 80
  },
  "draft": false,
  "categories": [
    "Java",
    "EE"
  ],
  "layout": "post",
  "author": "main",
  "disqus": true,
  "navigation": true,
  "current": "post",
  "tags": [
    "Object-Oriented Programming",
    "OOP",
    "Spring"
  ],
  "picture": {
    "src": "assets/resources/oop-within-enterprise/pexels-photo-577585.jpeg",
    "width": 1280,
    "height": 959,
    "alt": "A pair of glasses on a background of screens displaying code"
  },
  "sourcecode": {
    "link": "https://github.com/nfrankel/oopspring",
    "format": "Maven"
  },
  "gofurther": [
    {
      "url": "https://www.martinfowler.com/bliki/AnemicDomainModel.html",
      "text": "Anemic Domain Model"
    },
    {
      "url": "https://docs.spring.io/spring/docs/current/spring-framework-reference/core.html#beans-stereotype-annotations",
      "text": "@Component and further stereotype annotations"
    },
    {
      "url": "https://docs.spring.io/spring/docs/current/spring-framework-reference/core.html#aop-using-aspectj",
      "text": "Using AspectJ to dependency inject domain objects with Spring"
    },
    {
      "url": "https://docs.spring.io/spring/docs/current/spring-framework-reference/core.html#aop-aj-ltw",
      "text": "Load-time weaving with AspectJ in the Spring Framework"
    },
    {
      "url": "http://olivergierke.de/2013/11/why-field-injection-is-evil/",
      "text": "Why field injection is evil"
    }
  ],
  "title": "Is Object-Oriented Programming compatible with an enteprise context?",
  "ext": ".adoc",
  "document": "#<Asciidoctor::Document@70289126102160 {doctype: \"article\", doctitle: nil, blocks: 4}>",
  "date": 1510441200,
  "excerpt_html": "<div id=\"preamble\"> <div class=\"sectionbody\"> <div class=\"paragraph\"> <p>This week, during a workshop related to a Java course I give at a higher education school, I noticed the code produced by the students was mostly - ok, entirely, procedural. In fact, though the Java language touts itself as an Object-Oriented language, it&#8217;s not uncommon to find such code developed by professional developers in enterprises. For example, <a href=\"https://en.wikipedia.org/wiki/JavaBeans\" target=\"_blank\" rel=\"noopener\">the JavaBean specification</a> is in direct contradiction of one of OOP&#8217;s main principle, <a href=\"https://en.wikipedia.org/wiki/Encapsulation_(computer_programming)\" target=\"_blank\" rel=\"noopener\">encapsulation</a>.</p> </div> <div class=\"paragraph\"> <p>Another example is the widespread controller, service and DAO architecture found equally in Java EE and Spring applications. In that context, entities are in general <a href=\"https://www.martinfowler.com/bliki/AnemicDomainModel.html\" target=\"_blank\" rel=\"noopener\">anemic</a>, while all business logic is located in the service layer. While this is not bad <em>per se</em>, this design separates between state and behaviour, and sits at the opposite to true OOP.</p> </div> <div class=\"paragraph\"> <p>Both Java EE and the Spring framework enforce this layered design. For example, in Spring, there&#8217;s one annotation for every such layer: <code>@Controller</code>, <code>@Service</code> and <code>@Repository</code>. In the Java EE world, only <code>@EJB</code> instances - the service layer, can be made transactional.</p> </div> <div class=\"paragraph\"> <p>This post aims to try to reconcile both the OOP paradigm and the layered architecture. I&#8217;ll be using the Spring framework to highlight my point because I&#8217;m more familiar with it, but I believe the same approach could be used for pure Java EE apps.</p> </div> </div> </div> <div class=\"sect1\"> <h2 id=\"a-simple-use-case\">A simple use-case</h2> <div class=\"sectionbody\"> <div class=\"paragraph\"> <p>Let&#8217;s have a simple use-case: from an IBAN number, find the associated account with the relevant balance. Within a standard design, this could look like that:</p> </div> <div class=\"listingblock\"> <div class=\"content\"> <pre class=\"highlight\"><code class=\"language-kotlin\" data-lang=\"kotlin\"><span id=\"L1\" class=\"line\"><span class=\"n\">@RestController</span></span> <span id=\"L2\" class=\"line\"><span class=\"kd\">class</span> <span class=\"nc\">ClassicAccountController</span><span class=\"p\">(</span><span class=\"k\">private</span> <span class=\"kd\">val</span> <span class=\"py\">service</span><span class=\"p\">:</span> <span class=\"n\">AccountService</span><span class=\"p\">)</span> <span class=\"p\">{</span></span> <span id=\"L3\" class=\"line\"></span> <span id=\"L4\" class=\"line\">    <span class=\"n\">@GetMapping</span><span class=\"p\">(</span><span class=\"s\">\"/classicaccount/{iban}\"</span><span class=\"p\">)</span></span> <span id=\"L5\" class=\"line\">    <span class=\"k\">fun</span> <span class=\"nf\">getAccount</span><span class=\"p\">(</span><span class=\"n\">@PathVariable</span><span class=\"p\">(</span><span class=\"s\">\"iban\"</span><span class=\"p\">)</span> <span class=\"n\">iban</span><span class=\"p\">:</span> <span class=\"n\">String</span><span class=\"p\">)</span> <span class=\"p\">=</span> <span class=\"n\">service</span><span class=\"p\">.</span><span class=\"n\">findAccount</span><span class=\"p\">(</span><span class=\"n\">iban</span><span class=\"p\">)</span></span> <span id=\"L6\" class=\"line\"><span class=\"p\">}</span></span> <span id=\"L7\" class=\"line\"></span> <span id=\"L8\" class=\"line\"><span class=\"n\">@Service</span></span> <span id=\"L9\" class=\"line\"><span class=\"kd\">class</span> <span class=\"nc\">AccountService</span><span class=\"p\">(</span><span class=\"k\">private</span> <span class=\"kd\">val</span> <span class=\"py\">repository</span><span class=\"p\">:</span> <span class=\"n\">ClassicAccountRepository</span><span class=\"p\">)</span> <span class=\"p\">{</span></span> <span id=\"L10\" class=\"line\">    <span class=\"k\">fun</span> <span class=\"nf\">findAccount</span><span class=\"p\">(</span><span class=\"n\">iban</span><span class=\"p\">:</span> <span class=\"n\">String</span><span class=\"p\">)</span> <span class=\"p\">=</span> <span class=\"n\">repository</span><span class=\"p\">.</span><span class=\"n\">findOne</span><span class=\"p\">(</span><span class=\"n\">iban</span><span class=\"p\">)</span></span> <span id=\"L11\" class=\"line\"><span class=\"p\">}</span></span> <span id=\"L12\" class=\"line\"></span> <span id=\"L13\" class=\"line\"><span class=\"kd\">interface</span> <span class=\"nc\">ClassicAccountRepository</span> <span class=\"p\">:</span> <span class=\"n\">CrudRepository</span><span class=\"p\">&lt;</span><span class=\"n\">ClassicAccount</span><span class=\"p\">,</span> <span class=\"n\">String</span><span class=\"p\">&gt;</span></span> <span id=\"L14\" class=\"line\"></span> <span id=\"L15\" class=\"line\"><span class=\"n\">@Entity</span></span> <span id=\"L16\" class=\"line\"><span class=\"n\">@Table</span><span class=\"p\">(</span><span class=\"n\">name</span> <span class=\"p\">=</span> <span class=\"s\">\"ACCOUNT\"</span><span class=\"p\">)</span></span> <span id=\"L17\" class=\"line\"><span class=\"kd\">class</span> <span class=\"nc\">ClassicAccount</span><span class=\"p\">(</span><span class=\"n\">@Id</span> <span class=\"kd\">var</span> <span class=\"py\">iban</span><span class=\"p\">:</span> <span class=\"n\">String</span> <span class=\"p\">=</span> <span class=\"s\">\"\"</span><span class=\"p\">,</span> <span class=\"kd\">var</span> <span class=\"py\">balance</span><span class=\"p\">:</span> <span class=\"n\">BigDecimal</span> <span class=\"p\">=</span> <span class=\"n\">BigDecimal</span><span class=\"p\">.</span><span class=\"n\">ZERO</span><span class=\"p\">)</span></span></code></pre> </div> </div> <div class=\"paragraph\"> <p>There are a couple of issues there:</p> </div> <ol class=\"arabic\"><li><span class=\"primary\">The JPA specifications mandates for a no-arg constructor. Hence, it&#8217;s possible to create <code>ClassicalAccount</code> instances with an empty IBAN.</span></li><li><span class=\"primary\">There&#8217;s no validation of the IBAN. The full round-trip to the database is required to check if an IBAN is valid.</span></li></ol> <div class=\"admonitionblock note\"> <table> <tr> <td class=\"icon\"> <i class=\"fa icon-note\" title=\"Note\"></i> </td> <td class=\"content\"> Yes, there&#8217;s no currency. It&#8217;s a simple example, remember? </td> </tr> </table> </div> </div> </div> <div class=\"sect1\"> <h2 id=\"being-compliant\">Being compliant</h2> <div class=\"sectionbody\"> <div class=\"paragraph\"> <p>In order to comply with the no-args constructor JPA constraint - and because we use Kotlin, it&#8217;s possible to generate a synthetic constructor. That means the constructor is accessible through reflection, but <strong>not</strong> by calling the constructor directly.</p> </div> <div class=\"listingblock\"> <div class=\"content\"> <pre class=\"highlight\"><code class=\"language-xml\" data-lang=\"xml\"><span id=\"L1\" class=\"line\"><span class=\"nt\">&lt;plugin&gt;</span></span> <span id=\"L2\" class=\"line\">    <span class=\"nt\">&lt;artifactId&gt;</span>kotlin-maven-plugin<span class=\"nt\">&lt;/artifactId&gt;</span></span> <span id=\"L3\" class=\"line\">    <span class=\"nt\">&lt;groupId&gt;</span>org.jetbrains.kotlin<span class=\"nt\">&lt;/groupId&gt;</span></span> <span id=\"L4\" class=\"line\">    <span class=\"nt\">&lt;version&gt;</span>${kotlin.version}<span class=\"nt\">&lt;/version&gt;</span></span> <span id=\"L5\" class=\"line\">    <span class=\"nt\">&lt;configuration&gt;</span></span> <span id=\"L6\" class=\"line\">        <span class=\"nt\">&lt;compilerPlugins&gt;</span></span> <span id=\"L7\" class=\"line\">            <span class=\"nt\">&lt;plugin&gt;</span>jpa<span class=\"nt\">&lt;/plugin&gt;</span></span> <span id=\"L8\" class=\"line\">        <span class=\"nt\">&lt;/compilerPlugins&gt;</span></span> <span id=\"L9\" class=\"line\">    <span class=\"nt\">&lt;/configuration&gt;</span></span> <span id=\"L10\" class=\"line\">    <span class=\"nt\">&lt;dependencies&gt;</span></span> <span id=\"L11\" class=\"line\">        <span class=\"nt\">&lt;dependency&gt;</span></span> <span id=\"L12\" class=\"line\">            <span class=\"nt\">&lt;groupId&gt;</span>org.jetbrains.kotlin<span class=\"nt\">&lt;/groupId&gt;</span></span> <span id=\"L13\" class=\"line\">            <span class=\"nt\">&lt;artifactId&gt;</span>kotlin-maven-noarg<span class=\"nt\">&lt;/artifactId&gt;</span></span> <span id=\"L14\" class=\"line\">            <span class=\"nt\">&lt;version&gt;</span>${kotlin.version}<span class=\"nt\">&lt;/version&gt;</span></span> <span id=\"L15\" class=\"line\">        <span class=\"nt\">&lt;/dependency&gt;</span></span> <span id=\"L16\" class=\"line\">    <span class=\"nt\">&lt;/dependencies&gt;</span></span> <span id=\"L17\" class=\"line\"><span class=\"nt\">&lt;/plugin&gt;</span></span></code></pre> </div> </div> <div class=\"admonitionblock warning\"> <table> <tr> <td class=\"icon\"> <i class=\"fa icon-warning\" title=\"Warning\"></i> </td> <td class=\"content\"> If you use Java, tough luck, I don&#8217;t know about any option to fix that. </td> </tr> </table> </div> </div> </div> <div class=\"sect1\"> <h2 id=\"adding-validation\">Adding validation</h2> <div class=\"sectionbody\"> <div class=\"paragraph\"> <p>In a layer architecture, the service layer is the obvious place to put the business logic, including validation:</p> </div> <div class=\"listingblock\"> <div class=\"content\"> <pre class=\"highlight\"><code class=\"language-kotlin\" data-lang=\"kotlin\"><span id=\"L1\" class=\"line\"><span class=\"n\">@Service</span></span> <span id=\"L2\" class=\"line\"><span class=\"kd\">class</span> <span class=\"nc\">AccountService</span><span class=\"p\">(</span><span class=\"k\">private</span> <span class=\"kd\">val</span> <span class=\"py\">repository</span><span class=\"p\">:</span> <span class=\"n\">ClassicAccountRepository</span><span class=\"p\">)</span> <span class=\"p\">{</span></span> <span id=\"L3\" class=\"line\">    <span class=\"k\">fun</span> <span class=\"nf\">findAccount</span><span class=\"p\">(</span><span class=\"n\">iban</span><span class=\"p\">:</span> <span class=\"n\">String</span><span class=\"p\">):</span> <span class=\"n\">Account</span><span class=\"p\">?</span> <span class=\"p\">{</span></span> <span id=\"L4\" class=\"line\">        <span class=\"n\">checkIban</span><span class=\"p\">(</span><span class=\"n\">iban</span><span class=\"p\">)</span></span> <span id=\"L5\" class=\"line\">        <span class=\"k\">return</span> <span class=\"n\">repository</span><span class=\"p\">.</span><span class=\"n\">findOne</span><span class=\"p\">(</span><span class=\"n\">iban</span><span class=\"p\">)</span></span> <span id=\"L6\" class=\"line\">    <span class=\"p\">}</span></span> <span id=\"L7\" class=\"line\"></span> <span id=\"L8\" class=\"line\">    <span class=\"k\">fun</span> <span class=\"nf\">checkIban</span><span class=\"p\">(</span><span class=\"n\">iban</span><span class=\"p\">:</span> <span class=\"n\">String</span><span class=\"p\">)</span> <span class=\"p\">{</span></span> <span id=\"L9\" class=\"line\">        <span class=\"k\">if</span> <span class=\"p\">(</span><span class=\"n\">iban</span><span class=\"p\">.</span><span class=\"n\">isBlank</span><span class=\"p\">())</span> <span class=\"k\">throw</span> <span class=\"n\">IllegalArgumentException</span><span class=\"p\">(</span><span class=\"s\">\"IBAN cannot be blank\"</span><span class=\"p\">)</span></span> <span id=\"L10\" class=\"line\">    <span class=\"p\">}</span></span> <span id=\"L11\" class=\"line\"><span class=\"p\">}</span></span></code></pre> </div> </div> <div class=\"paragraph\"> <p>In order to be more OOP-compliant, we must decide whether we should allow invalid IBAN numbers or not. It&#8217;s easier to forbid it altogether.</p> </div> <div class=\"listingblock\"> <div class=\"content\"> <pre class=\"highlight\"><code class=\"language-kotlin\" data-lang=\"kotlin\"><span id=\"L1\" class=\"line\"><span class=\"n\">@Entity</span></span> <span id=\"L2\" class=\"line\"><span class=\"n\">@Table</span><span class=\"p\">(</span><span class=\"n\">name</span> <span class=\"p\">=</span> <span class=\"s\">\"ACCOUNT\"</span><span class=\"p\">)</span></span> <span id=\"L3\" class=\"line\"><span class=\"kd\">class</span> <span class=\"nc\">OopAccount</span><span class=\"p\">(</span><span class=\"n\">@Id</span> <span class=\"kd\">var</span> <span class=\"py\">iban</span><span class=\"p\">:</span> <span class=\"n\">String</span><span class=\"p\">,</span> <span class=\"kd\">var</span> <span class=\"py\">balance</span><span class=\"p\">:</span> <span class=\"n\">BigDecimal</span> <span class=\"p\">=</span> <span class=\"n\">BigDecimal</span><span class=\"p\">.</span><span class=\"n\">ZERO</span><span class=\"p\">)</span> <span class=\"p\">{</span></span> <span id=\"L4\" class=\"line\">    <span class=\"n\">init</span> <span class=\"p\">{</span></span> <span id=\"L5\" class=\"line\">        <span class=\"k\">if</span> <span class=\"p\">(</span><span class=\"n\">iban</span><span class=\"p\">.</span><span class=\"n\">isBlank</span><span class=\"p\">())</span> <span class=\"k\">throw</span> <span class=\"n\">IllegalArgumentException</span><span class=\"p\">(</span><span class=\"s\">\"IBAN cannot be blank\"</span><span class=\"p\">)</span></span> <span id=\"L6\" class=\"line\">    <span class=\"p\">}</span></span> <span id=\"L7\" class=\"line\"><span class=\"p\">}</span></span></code></pre> </div> </div> <div class=\"paragraph\"> <p>However, this means that we must first create the <code>OopAccount</code> instance to validate the IBAN - with a balance of 0, even if the balance is actually not 0. Again, as per the empty IBAN, the code does not match the model. Even worse, to use the repository we must access the <code>OopAccount</code> inner state:</p> </div> <div class=\"listingblock\"> <div class=\"content\"> <pre class=\"highlight\"><code class=\"language-kotlin\" data-lang=\"kotlin\"><span id=\"L1\" class=\"line\"><span class=\"n\">repository</span><span class=\"p\">.</span><span class=\"n\">findOne</span><span class=\"p\">(</span><span class=\"n\">OopAccount</span><span class=\"p\">(</span><span class=\"n\">iban</span><span class=\"p\">).</span><span class=\"n\">iban</span><span class=\"p\">)</span></span></code></pre> </div> </div> </div> </div> <div class=\"sect1\"> <h2 id=\"a-more-oop-friendly-design\">A more OOP-friendly design</h2> <div class=\"sectionbody\"> <div class=\"paragraph\"> <p>Improving the state of the code requires a major rework on the class model, separating between the IBAN and the account, so that the former can be validated, and can access the latter. The IBAN class serves both as the entry point, and the PK of the account.</p> </div> <div class=\"listingblock\"> <div class=\"content\"> <pre class=\"highlight\"><code class=\"language-kotlin\" data-lang=\"kotlin\"><span id=\"L1\" class=\"line\"><span class=\"n\">@Entity</span></span> <span id=\"L2\" class=\"line\"><span class=\"n\">@Table</span><span class=\"p\">(</span><span class=\"n\">name</span> <span class=\"p\">=</span> <span class=\"s\">\"ACCOUNT\"</span><span class=\"p\">)</span></span> <span id=\"L3\" class=\"line\"><span class=\"kd\">class</span> <span class=\"nc\">OopAccount</span><span class=\"p\">(</span><span class=\"n\">@EmbeddedId</span> <span class=\"kd\">var</span> <span class=\"py\">iban</span><span class=\"p\">:</span> <span class=\"n\">Iban</span><span class=\"p\">,</span> <span class=\"kd\">var</span> <span class=\"py\">balance</span><span class=\"p\">:</span> <span class=\"n\">BigDecimal</span><span class=\"p\">)</span></span> <span id=\"L4\" class=\"line\"></span> <span id=\"L5\" class=\"line\"><span class=\"kd\">class</span> <span class=\"nc\">Iban</span><span class=\"p\">(</span><span class=\"n\">@Column</span><span class=\"p\">(</span><span class=\"n\">name</span> <span class=\"p\">=</span> <span class=\"s\">\"iban\"</span><span class=\"p\">)</span> <span class=\"kd\">val</span> <span class=\"py\">number</span><span class=\"p\">:</span> <span class=\"n\">String</span><span class=\"p\">,</span></span> <span id=\"L6\" class=\"line\">           <span class=\"n\">@Transient</span> <span class=\"k\">private</span> <span class=\"kd\">val</span> <span class=\"py\">repository</span><span class=\"p\">:</span> <span class=\"n\">OopAccountRepository</span><span class=\"p\">)</span> <span class=\"p\">:</span> <span class=\"n\">Serializable</span> <span class=\"p\">{</span></span> <span id=\"L7\" class=\"line\"></span> <span id=\"L8\" class=\"line\">    <span class=\"n\">init</span> <span class=\"p\">{</span></span> <span id=\"L9\" class=\"line\">        <span class=\"k\">if</span> <span class=\"p\">(</span><span class=\"n\">number</span><span class=\"p\">.</span><span class=\"n\">isBlank</span><span class=\"p\">())</span> <span class=\"k\">throw</span> <span class=\"n\">IllegalArgumentException</span><span class=\"p\">(</span><span class=\"s\">\"IBAN cannot be blank\"</span><span class=\"p\">)</span></span> <span id=\"L10\" class=\"line\">    <span class=\"p\">}</span></span> <span id=\"L11\" class=\"line\"></span> <span id=\"L12\" class=\"line\">    <span class=\"kd\">val</span> <span class=\"py\">account</span></span> <span id=\"L13\" class=\"line\">        <span class=\"n\">@JsonIgnore</span></span> <span id=\"L14\" class=\"line\">        <span class=\"k\">get</span><span class=\"p\">()</span> <span class=\"p\">=</span> <span class=\"n\">repository</span><span class=\"p\">.</span><span class=\"n\">findOne</span><span class=\"p\">(</span><span class=\"k\">this</span><span class=\"p\">)</span></span> <span id=\"L15\" class=\"line\"><span class=\"p\">}</span></span></code></pre> </div> </div> <div class=\"admonitionblock note\"> <table> <tr> <td class=\"icon\"> <i class=\"fa icon-note\" title=\"Note\"></i> </td> <td class=\"content\"> Notice the returned JSON structure will be different from the one returned above. If that&#8217;s an issue, it&#8217;s quite easy to customize Jackson to obtain the desired result. </td> </tr> </table> </div> <div class=\"paragraph\"> <p>With this new design, the controller requires a bit of change:</p> </div> <div class=\"listingblock\"> <div class=\"content\"> <pre class=\"highlight\"><code class=\"language-kotlin\" data-lang=\"kotlin\"><span id=\"L1\" class=\"line\"><span class=\"n\">@RestController</span></span> <span id=\"L2\" class=\"line\"><span class=\"kd\">class</span> <span class=\"nc\">OopAccountController</span><span class=\"p\">(</span><span class=\"k\">private</span> <span class=\"kd\">val</span> <span class=\"py\">repository</span><span class=\"p\">:</span> <span class=\"n\">OopAccountRepository</span><span class=\"p\">)</span> <span class=\"p\">{</span></span> <span id=\"L3\" class=\"line\"></span> <span id=\"L4\" class=\"line\">    <span class=\"n\">@GetMapping</span><span class=\"p\">(</span><span class=\"s\">\"/oopaccount/{iban}\"</span><span class=\"p\">)</span></span> <span id=\"L5\" class=\"line\">    <span class=\"k\">fun</span> <span class=\"nf\">getAccount</span><span class=\"p\">(</span><span class=\"n\">@PathVariable</span><span class=\"p\">(</span><span class=\"s\">\"iban\"</span><span class=\"p\">)</span> <span class=\"n\">number</span><span class=\"p\">:</span> <span class=\"n\">String</span><span class=\"p\">):</span> <span class=\"n\">OopAccount</span> <span class=\"p\">{</span></span> <span id=\"L6\" class=\"line\">        <span class=\"kd\">val</span> <span class=\"py\">iban</span> <span class=\"p\">=</span> <span class=\"n\">Iban</span><span class=\"p\">(</span><span class=\"n\">number</span><span class=\"p\">,</span> <span class=\"n\">repository</span><span class=\"p\">)</span></span> <span id=\"L7\" class=\"line\">        <span class=\"k\">return</span> <span class=\"n\">iban</span><span class=\"p\">.</span><span class=\"n\">account</span></span> <span id=\"L8\" class=\"line\">    <span class=\"p\">}</span></span> <span id=\"L9\" class=\"line\"><span class=\"p\">}</span></span></code></pre> </div> </div> <div class=\"paragraph\"> <p>The only disadvantage of this approach is that the repository needs to be injected into the controller, then be explicitly passed to the entity&#8217;s constructor.</p> </div> </div> </div> <div class=\"sect1\"> <h2 id=\"the-final-touch\">The final touch</h2> <div class=\"sectionbody\"> <div class=\"paragraph\"> <p>It would be great if the repository could automatically be injected into the entity when it&#8217;s created. Well, Spring makes it possible - though this is not a very well-known feature, through Aspect-Oriented Programming. It requires the following steps:</p> </div> <div class=\"dlist\"> <dl> <dt class=\"hdlist1\">Add AOP capabilities to the application</dt> <dd> <p>To effectively add the AOP dependency is quite straightforward and requires just adding the relevant starter dependency to the POM:</p> <div class=\"listingblock\"> <div class=\"content\"> <pre class=\"highlight\"><code class=\"language-xml\" data-lang=\"xml\"><span id=\"L1\" class=\"line\"><span class=\"nt\">&lt;dependency&gt;</span></span> <span id=\"L2\" class=\"line\">    <span class=\"nt\">&lt;groupId&gt;</span>org.springframework.boot<span class=\"nt\">&lt;/groupId&gt;</span></span> <span id=\"L3\" class=\"line\">    <span class=\"nt\">&lt;artifactId&gt;</span>spring-boot-starter-aop<span class=\"nt\">&lt;/artifactId&gt;</span></span> <span id=\"L4\" class=\"line\"><span class=\"nt\">&lt;/dependency&gt;</span></span></code></pre> </div> </div> <div class=\"paragraph\"> <p>Then, the application must be configured to make use of it:</p> </div> <div class=\"listingblock\"> <div class=\"content\"> <pre class=\"highlight\"><code class=\"language-kotlin\" data-lang=\"kotlin\"><span id=\"L1\" class=\"line\"><span class=\"n\">@SpringBootApplication</span></span> <span id=\"L2\" class=\"line\"><span class=\"n\">@EnableSpringConfigured</span></span> <span id=\"L3\" class=\"line\"><span class=\"kd\">class</span> <span class=\"nc\">OopspringApplication</span></span></code></pre> </div> </div> </dd> <dt class=\"hdlist1\">Update the entity</dt> <dd> <ol class=\"arabic\"><li><span class=\"primary\">The entity must first be set as a target for injection. Dependency injection will be done through autowiring.</span></li><li><span class=\"primary\">Then, the repository be moved from a constructor argument to a field.</span></li><li><span class=\"primary\">Finally, the database fetching logic can be moved into the entity:</span><div class=\"listingblock\"> <div class=\"content\"> <pre class=\"highlight\"><code class=\"language-kotlin\" data-lang=\"kotlin\"><span id=\"L1\" class=\"line\"><span class=\"n\">@Configurable</span><span class=\"p\">(</span><span class=\"n\">autowire</span> <span class=\"p\">=</span> <span class=\"n\">Autowire</span><span class=\"p\">.</span><span class=\"n\">BY_TYPE</span><span class=\"p\">)</span></span> <span id=\"L2\" class=\"line\"><span class=\"kd\">class</span> <span class=\"nc\">Iban</span><span class=\"p\">(</span><span class=\"n\">@Column</span><span class=\"p\">(</span><span class=\"n\">name</span> <span class=\"p\">=</span> <span class=\"s\">\"iban\"</span><span class=\"p\">)</span> <span class=\"kd\">val</span> <span class=\"py\">number</span><span class=\"p\">:</span> <span class=\"n\">String</span><span class=\"p\">)</span> <span class=\"p\">:</span> <span class=\"n\">Serializable</span> <span class=\"p\">{</span></span> <span id=\"L3\" class=\"line\"></span> <span id=\"L4\" class=\"line\">    <span class=\"n\">@Transient</span></span> <span id=\"L5\" class=\"line\">    <span class=\"n\">@Autowired</span></span> <span id=\"L6\" class=\"line\">    <span class=\"k\">private</span> <span class=\"k\">lateinit</span> <span class=\"kd\">var</span> <span class=\"py\">repository</span><span class=\"p\">:</span> <span class=\"n\">OopAccountRepository</span></span> <span id=\"L7\" class=\"line\"></span> <span id=\"L8\" class=\"line\">    <span class=\"n\">init</span> <span class=\"p\">{</span></span> <span id=\"L9\" class=\"line\">        <span class=\"k\">if</span> <span class=\"p\">(</span><span class=\"n\">number</span><span class=\"p\">.</span><span class=\"n\">isBlank</span><span class=\"p\">())</span> <span class=\"k\">throw</span> <span class=\"n\">IllegalArgumentException</span><span class=\"p\">(</span><span class=\"s\">\"IBAN cannot be blank\"</span><span class=\"p\">)</span></span> <span id=\"L10\" class=\"line\">    <span class=\"p\">}</span></span> <span id=\"L11\" class=\"line\"></span> <span id=\"L12\" class=\"line\">    <span class=\"kd\">val</span> <span class=\"py\">account</span></span> <span id=\"L13\" class=\"line\">        <span class=\"n\">@JsonIgnore</span></span> <span id=\"L14\" class=\"line\">        <span class=\"k\">get</span><span class=\"p\">()</span> <span class=\"p\">=</span> <span class=\"n\">repository</span><span class=\"p\">.</span><span class=\"n\">findOne</span><span class=\"p\">(</span><span class=\"k\">this</span><span class=\"p\">)</span></span> <span id=\"L15\" class=\"line\"><span class=\"p\">}</span></span></code></pre> </div> </div></li></ol> </dd> </dl> </div> <div class=\"admonitionblock note\"> <table> <tr> <td class=\"icon\"> <i class=\"fa icon-note\" title=\"Note\"></i> </td> <td class=\"content\"> Remember that field-injection is evil. </td> </tr> </table> </div> <div class=\"dlist\"> <dl> <dt class=\"hdlist1\">Aspect weaving</dt> <dd> <p>There are two ways to weave the aspect into the , either compile-time weaving, or load-time weaving. I choose the the later is much easier to configure. It&#8217;s achieved through a standard Java agent.</p> <ol class=\"arabic\"><li><span class=\"primary\">First, it needs to be added as a runtime dependency in the POM:</span><div class=\"listingblock\"> <div class=\"content\"> <pre class=\"highlight\"><code class=\"language-xml\" data-lang=\"xml\"><span id=\"L1\" class=\"line\"><span class=\"nt\">&lt;dependency&gt;</span></span> <span id=\"L2\" class=\"line\">    <span class=\"nt\">&lt;groupId&gt;</span>org.springframework<span class=\"nt\">&lt;/groupId&gt;</span></span> <span id=\"L3\" class=\"line\">    <span class=\"nt\">&lt;artifactId&gt;</span>spring-agent<span class=\"nt\">&lt;/artifactId&gt;</span></span> <span id=\"L4\" class=\"line\">    <span class=\"nt\">&lt;version&gt;</span>2.5.6<span class=\"nt\">&lt;/version&gt;</span></span> <span id=\"L5\" class=\"line\">    <span class=\"nt\">&lt;scope&gt;</span>runtime<span class=\"nt\">&lt;/scope&gt;</span></span> <span id=\"L6\" class=\"line\"><span class=\"nt\">&lt;/dependency&gt;</span></span></code></pre> </div> </div></li><li><span class=\"primary\">Then, the Spring Boot plugin must be configured with the agent:</span><div class=\"listingblock\"> <div class=\"content\"> <pre class=\"highlight\"><code class=\"language-xml\" data-lang=\"xml\"><span id=\"L1\" class=\"line\"><span class=\"nt\">&lt;plugin&gt;</span></span> <span id=\"L2\" class=\"line\">    <span class=\"nt\">&lt;groupId&gt;</span>org.springframework.boot<span class=\"nt\">&lt;/groupId&gt;</span></span> <span id=\"L3\" class=\"line\">    <span class=\"nt\">&lt;artifactId&gt;</span>spring-boot-maven-plugin<span class=\"nt\">&lt;/artifactId&gt;</span></span> <span id=\"L4\" class=\"line\">    <span class=\"nt\">&lt;configuration&gt;</span></span> <span id=\"L5\" class=\"line\">        <span class=\"nt\">&lt;agent&gt;</span>${settings.localRepository}/org/springframework/spring-agent/2.5.6/spring-agent-2.5.6.jar<span class=\"nt\">&lt;/agent&gt;</span></span> <span id=\"L6\" class=\"line\">    <span class=\"nt\">&lt;/configuration&gt;</span></span> <span id=\"L7\" class=\"line\"><span class=\"nt\">&lt;/plugin&gt;</span></span></code></pre> </div> </div></li><li><span class=\"primary\">Finally, the application must be configured accordingly:</span><div class=\"listingblock\"> <div class=\"content\"> <pre class=\"highlight\"><code class=\"language-kotlin\" data-lang=\"kotlin\"><span id=\"L1\" class=\"line\"><span class=\"n\">@EnableLoadTimeWeaving</span></span> <span id=\"L2\" class=\"line\"><span class=\"kd\">class</span> <span class=\"nc\">OopspringApplication</span></span></code></pre> </div> </div></li></ol> </dd> </dl> </div> </div> </div> <div class=\"sect1\"> <h2 id=\"and-then\">And then?</h2> <div class=\"sectionbody\"> <div class=\"paragraph\"> <p>Of course, this sample leaves out an important piece of the design: how to update the balance of an account? The layer approach would have a setter for that, but that&#8217;s not OOP. Thinking about it, the balance of an account changes because there&#8217;s a transfer from another account. This could be modeled as:</p> </div> <div class=\"listingblock\"> <div class=\"content\"> <pre class=\"highlight\"><code class=\"language-kotlin\" data-lang=\"kotlin\"><span id=\"L1\" class=\"line\"><span class=\"k\">fun</span> <span class=\"nf\">OopAccount</span><span class=\"p\">.</span><span class=\"n\">transfer</span><span class=\"p\">(</span><span class=\"n\">source</span><span class=\"p\">:</span> <span class=\"n\">OopAccount</span><span class=\"p\">,</span> <span class=\"n\">amount</span><span class=\"p\">:</span> <span class=\"n\">BigDecimal</span><span class=\"p\">)</span> <span class=\"p\">{</span> <span class=\"o\">..</span><span class=\"p\">.</span> <span class=\"p\">}</span></span></code></pre> </div> </div> <div class=\"paragraph\"> <p>Experienced developers should see some transaction management requirements sneaking in. I leave the implementation to motivated readers. The next step would be to cache the values, because accessing the database for each read and write would be killing performance.</p> </div> </div> </div> <div class=\"sect1\"> <h2 id=\"conclusion\">Conclusion</h2> <div class=\"sectionbody\"> <div class=\"paragraph\"> <p>There are a couple of points I&#8217;d like to make.</p> </div> <div class=\"paragraph\"> <p>First, the answer to the title question is a resounding 'yes'. The results is a true OOP code while still using a so-called enterprise-grade framework - namely Spring.</p> </div> <div class=\"paragraph\"> <p>However, migrating to this OOP-compatible design came with a bit of overhead. Not only did we rely on field injection, we had to bring in AOP with load-time weaving. The first is a hindrance during unit testing, the second is a technology you definitely don&#8217;t want in every team, as they make apps more complex. And that&#8217;s only for a trivial example.</p> </div> <div class=\"paragraph\"> <p>Finally, this approach has a huge drawback: most developers are not familiar with it. Whatever its advantages, they first must be \"conditioned\" to have this mindset. And that might be a reason to continue using the traditional layered architecture.</p> </div> <div class=\"admonitionblock note\"> <table> <tr> <td class=\"icon\"> <i class=\"fa icon-note\" title=\"Note\"></i> </td> <td class=\"content\"> I&#8217;ve searched Google for <strong>scientific</strong> studies proving OOP is better in terms of readability and maintainability: I found none. I would be very grateful for pointers. </td> </tr> </table> </div> </div> </div>",
  "excerpt_text": "This week, during a workshop related to a Java course I give at a higher education school, I noticed the code produced by the students was mostly - ok, entirely, procedural. In fact, though the Java language touts itself as an Object-Oriented language, it’s not uncommon to find such code developed by professional developers in enterprises. For example, the JavaBean specification is in direct contradiction of one of OOP’s main principle, encapsulation. Another example is the widespread controller, service and DAO architecture found equally in Java EE and Spring applications. In that context, entities are in general anemic, while all business logic is located in the service layer. While this is not bad per se, this design separates between state and behaviour, and sits at the opposite to true OOP. Both Java EE and the Spring framework enforce this layered design. For example, in Spring, there’s one annotation for every such layer: @Controller, @Service and @Repository. In the Java EE world, only @EJB instances - the service layer, can be made transactional. This post aims to try to reconcile both the OOP paradigm and the layered architecture. I’ll be using the Spring framework to highlight my point because I’m more familiar with it, but I believe the same approach could be used for pure Java EE apps. A simple use-case Let’s have a simple use-case: from an IBAN number, find the associated account with the relevant balance. Within a standard design, this could look like that: @RestController class ClassicAccountController(private val service: AccountService) { @GetMapping(\"/classicaccount/{iban}\") fun getAccount(@PathVariable(\"iban\") iban: String) = service.findAccount(iban) } @Service class AccountService(private val repository: ClassicAccountRepository) { fun findAccount(iban: String) = repository.findOne(iban) } interface ClassicAccountRepository : CrudRepository<ClassicAccount, String> @Entity @Table(name = \"ACCOUNT\") class ClassicAccount(@Id var iban: String = \"\", var balance: BigDecimal = BigDecimal.ZERO) There are a couple of issues there: The JPA specifications mandates for a no-arg constructor. Hence, it’s possible to create ClassicalAccount instances with an empty IBAN.There’s no validation of the IBAN. The full round-trip to the database is required to check if an IBAN is valid. Yes, there’s no currency. It’s a simple example, remember? Being compliant In order to comply with the no-args constructor JPA constraint - and because we use Kotlin, it’s possible to generate a synthetic constructor. That means the constructor is accessible through reflection, but not by calling the constructor directly. <plugin> <artifactId>kotlin-maven-plugin</artifactId> <groupId>org.jetbrains.kotlin</groupId> <version>${kotlin.version}</version> <configuration> <compilerPlugins> <plugin>jpa</plugin> </compilerPlugins> </configuration> <dependencies> <dependency> <groupId>org.jetbrains.kotlin</groupId> <artifactId>kotlin-maven-noarg</artifactId> <version>${kotlin.version}</version> </dependency> </dependencies> </plugin> If you use Java, tough luck, I don’t know about any option to fix that. Adding validation In a layer architecture, the service layer is the obvious place to put the business logic, including validation: @Service class AccountService(private val repository: ClassicAccountRepository) { fun findAccount(iban: String): Account? { checkIban(iban) return repository.findOne(iban) } fun checkIban(iban: String) { if (iban.isBlank()) throw IllegalArgumentException(\"IBAN cannot be blank\") } } In order to be more OOP-compliant, we must decide whether we should allow invalid IBAN numbers or not. It’s easier to forbid it altogether. @Entity @Table(name = \"ACCOUNT\") class OopAccount(@Id var iban: String, var balance: BigDecimal = BigDecimal.ZERO) { init { if (iban.isBlank()) throw IllegalArgumentException(\"IBAN cannot be blank\") } } However, this means that we must first create the OopAccount instance to validate the IBAN - with a balance of 0, even if the balance is actually not 0. Again, as per the empty IBAN, the code does not match the model. Even worse, to use the repository we must access the OopAccount inner state: repository.findOne(OopAccount(iban).iban) A more OOP-friendly design Improving the state of the code requires a major rework on the class model, separating between the IBAN and the account, so that the former can be validated, and can access the latter. The IBAN class serves both as the entry point, and the PK of the account. @Entity @Table(name = \"ACCOUNT\") class OopAccount(@EmbeddedId var iban: Iban, var balance: BigDecimal) class Iban(@Column(name = \"iban\") val number: String, @Transient private val repository: OopAccountRepository) : Serializable { init { if (number.isBlank()) throw IllegalArgumentException(\"IBAN cannot be blank\") } val account @JsonIgnore get() = repository.findOne(this) } Notice the returned JSON structure will be different from the one returned above. If that’s an issue, it’s quite easy to customize Jackson to obtain the desired result. With this new design, the controller requires a bit of change: @RestController class OopAccountController(private val repository: OopAccountRepository) { @GetMapping(\"/oopaccount/{iban}\") fun getAccount(@PathVariable(\"iban\") number: String): OopAccount { val iban = Iban(number, repository) return iban.account } } The only disadvantage of this approach is that the repository needs to be injected into the controller, then be explicitly passed to the entity’s constructor. The final touch It would be great if the repository could automatically be injected into the entity when it’s created. Well, Spring makes it possible - though this is not a very well-known feature, through Aspect-Oriented Programming. It requires the following steps: Add AOP capabilities to the application To effectively add the AOP dependency is quite straightforward and requires just adding the relevant starter dependency to the POM: <dependency> <groupId>org.springframework.boot</groupId> <artifactId>spring-boot-starter-aop</artifactId> </dependency> Then, the application must be configured to make use of it: @SpringBootApplication @EnableSpringConfigured class OopspringApplication Update the entity The entity must first be set as a target for injection. Dependency injection will be done through autowiring.Then, the repository be moved from a constructor argument to a field.Finally, the database fetching logic can be moved into the entity: @Configurable(autowire = Autowire.BY_TYPE) class Iban(@Column(name = \"iban\") val number: String) : Serializable { @Transient @Autowired private lateinit var repository: OopAccountRepository init { if (number.isBlank()) throw IllegalArgumentException(\"IBAN cannot be blank\") } val account @JsonIgnore get() = repository.findOne(this) } Remember that field-injection is evil. Aspect weaving There are two ways to weave the aspect into the , either compile-time weaving, or load-time weaving. I choose the the later is much easier to configure. It’s achieved through a standard Java agent. First, it needs to be added as a runtime dependency in the POM: <dependency> <groupId>org.springframework</groupId> <artifactId>spring-agent</artifactId> <version>2.5.6</version> <scope>runtime</scope> </dependency> Then, the Spring Boot plugin must be configured with the agent: <plugin> <groupId>org.springframework.boot</groupId> <artifactId>spring-boot-maven-plugin</artifactId> <configuration> <agent>${settings.localRepository}/org/springframework/spring-agent/2.5.6/spring-agent-2.5.6.jar</agent> </configuration> </plugin> Finally, the application must be configured accordingly: @EnableLoadTimeWeaving class OopspringApplication And then? Of course, this sample leaves out an important piece of the design: how to update the balance of an account? The layer approach would have a setter for that, but that’s not OOP. Thinking about it, the balance of an account changes because there’s a transfer from another account. This could be modeled as: fun OopAccount.transfer(source: OopAccount, amount: BigDecimal) { ... } Experienced developers should see some transaction management requirements sneaking in. I leave the implementation to motivated readers. The next step would be to cache the values, because accessing the database for each read and write would be killing performance. Conclusion There are a couple of points I’d like to make. First, the answer to the title question is a resounding 'yes'. The results is a true OOP code while still using a so-called enterprise-grade framework - namely Spring. However, migrating to this OOP-compatible design came with a bit of overhead. Not only did we rely on field injection, we had to bring in AOP with load-time weaving. The first is a hindrance during unit testing, the second is a technology you definitely don’t want in every team, as they make apps more complex. And that’s only for a trivial example. Finally, this approach has a huge drawback: most developers are not familiar with it. Whatever its advantages, they first must be \"conditioned\" to have this mindset. And that might be a reason to continue using the traditional layered architecture. I’ve searched Google for scientific studies proving OOP is better in terms of readability and maintainability: I found none. I would be very grateful for pointers.",
  "slug": "oop-compatible-enterprise-context",
  "type": "post",
  "url": "/oop-compatible-enterprise-context/",
  "objectID": "806c94300aed575014de5221b98d70ac"
}
```

**Comment by jdwallace12** [comment]
@pixelastic That did the trick! Thanks

**Comment by nfrankel** [comment]
FYI, I wrote a basic plugin (I'm not a Ruby developer) instead of a hook to cut on the excerpt size:

```ruby
Jekyll::Hooks.register :posts, :pre_render do |doc|
  content = doc.content
  if content.length > 150
    content = content[0..150]
  end
  doc.data['excerpt'] = content
end
```

Not only does indexing works now, it's also very very fast.

**Comment by pixelastic** [comment]
@nfrankel: Glad you found a workaround!

I release v1.1.2 that should fix the "double posting" of records when using `jekyll-asciidoc`. Could you confirm that when you run `jekyll algolia` twice in a row, the second call actually does nothing (identifying that there is nothing to update).

I've also understood why I was able to push your record and you weren't. The 10Kb limit on records is actually only applied to newer accounts. My old account (where I'm doing test) had a larger limit, allowing me to push. Now that you've found a way to reduce the excerpt and know exactly which record causes trouble, you should be good to go.

One last remark though. I see in your record that the `html` key contains a very large blurb of text. This is because you defined `dd` as a selector for your records. This might be too large a selector, and maybe `dl` might be enough for you here.

Anyway, I'm glad we managed to fix everything. I'm going to close this :)

**Comment by nfrankel** [comment]
I cannot test right now but just a comment about the remove/create behavior. I don't know how you implemented it, but I suggest you create a hash of the record. If the record in the index has the same hash, don't do anything, if not, then update. And of course, if the record is not found, remove it from the index.

**Comment by nfrankel** [comment]
I tested and got an issue I didn't have before. For whatever reason, now records are being _added_ in addition to the existing records which makes my index go over the limit... Will downgrade back.

**Comment by pixelastic** [comment]
> I don't know how you implemented it, but I suggest you create a hash of the record. If the record in the index has the same hash, don't do anything, if not, then update. And of course, if the record is not found, remove it from the index.

This is exactly how it works ;). The record `objectID` is a hash of the record content.

As for your second error, it seems you're having the same error as this one: https://github.com/algolia/jekyll-algolia/issues/39. Sorry for the inconvenience, I'm going to have a look.

**Comment by nfrankel** [comment]
Don't be sorry: you unblocked me in a very short timeframe. So thank you :-)

**Comment by pixelastic** [comment]
@nfrankel This should be fixed with v1.1.3. I tested with your `feature/search` branch and it works fine for me. Let me know if it fixes your issue.
